storage-proofs/porep/src/stacked/vanilla/proof.rs

```diff
@@ -497,28 +510,78 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher> StackedDrg<'a, Tr
                         column_tree_builder.add_columns(&columns)?;
                     } else {
                         assert_eq!(node_index, nodes_count);
-                        let tree_data = column_tree_builder.add_final_columns(&columns)?;
-                        let tree_data_len = tree_data.len();
+                        let (base_data, tree_data) =
+                            column_tree_builder.add_final_columns(&columns)?;
+                        trace!(
+                            "base data len {}, tree data len {}",
+                            base_data.len(),
+                            tree_data.len()
+                        );
+                        let tree_len = base_data.len() + tree_data.len();
                         info!(
                             "persisting base tree_c {}/{} of length {}",
                             i + 1,
                             tree_count,
-                            tree_data_len
+                            tree_len,
+                        );
+                        assert_eq!(base_data.len(), nodes_count);
+                        assert_eq!(tree_len, config.size.unwrap());
+
+                        // Persist the base and tree data to disk based using the current store config.
+                        let tree_c_store =
+                            DiskStore::<<Tree::Hasher as Hasher>::Domain>::new_with_config(
+                                tree_len,
+                                Tree::Arity::to_usize(),
+                                config.clone(),
+                            )?;
+
+                        let store = Arc::new(RwLock::new(tree_c_store));
+
+                        let base_offset = base_data.len();
+                        let batch_size = std::cmp::min(base_data.len(), column_write_batch_size);
+                        trace!(
+                            "flattening tree_c base data of {} nodes using batch size {}",
+                            base_data.len(),
+                            batch_size
                         );
-                        assert_eq!(tree_data_len, config.size.unwrap());
+                        base_data
+                            .into_par_iter()
+                            .chunks(column_write_batch_size)
+                            .enumerate()
+                            .try_for_each(|(index, fr_elements)| {
+                                let mut buf = Vec::with_capacity(batch_size * NODE_SIZE);
 
-                        let flat_tree_data: Vec<_> = tree_data
+                                for fr in fr_elements {
+                                    buf.extend(fr_into_bytes(&fr));
```

Comment:
This will still allocate a buffer into which to write the conversion. I think you should add a new version of this function which writes directly into `buf`.

Reply:
For this instance (of tree data), it's taking about a half second on a 32GiB run.  The base data flattening is now taking about 2-3 seconds.

Reply:
The buf being extended has been pre-allocated above that.  I'm not sure if you're saying to avoid that allocation, or the one that extend would do had it not been pre-allocated.

Reply:
I was talking about the allocation inside `fr_into_bytes`.

Reply:
:+1: Got it, thanks.  Given the overall speed improvements, we can wait on that a bit since this is no longer bottlenecking.